File: app/helpers/dropdown/participation_lists.rb
#  Copyright (c) 2018, Schweizer Blasmusikverband. This file is part of
#  hitobito and licensed under the Affero General Public License version 3
#  or later. See the COPYING file at the top-level directory or at
#  https://github.com/hitobito/hitobito.

module Dropdown
  class ParticipationLists < Dropdown::Base
    def initialize(template, group, translation)
      super(template, translation, :plus)
      @template = template
      @group = group
      init_items
    end

    private

    def init_items
      ::Event.all_types.each do |event|
        next unless authorized?(event)
        add_item(event.label,
          build_event_participation_lists_path(event),
          participation_lists_options)
      end
    end

    def authorized?(event)
      @template.can?(:"index_#{event.to_s.underscore.pluralize}", @group)
    end

    def build_event_participation_lists_path(event)
      type = event == ::Event ? nil : event.to_s
      template.group_events_participation_lists_new_path(@group, type: type, label: event.label)
    end

    def participation_lists_options
      {data: {checkable: true, method: :get}, remote: true}
    end
  end
end

File: app/javascript/javascripts/modules/participation_lists.js
//  Copyright (c) 2012-2018, Schweizer Blasmusikverband. This file is part of
//  hitobito_sbv and licensed under the Affero General Public License version 3
//  or later. See the COPYING file at the top-level directory or at
//  https://github.com/hitobito/hitobito.
var app;

app = window.App || (window.App = {});

app.ParticipationLists = {
  updatePath: function(e) {
    var event, form, metaToken;
    event = JSON.parse(e);
    form = $('form#new_event_participation')[0];
    $("#new_event_participation button[type='submit']").prop('disabled', false);
    form.action = form.action.replace(/(\d|-)*?(?=\/\w*$)/, event['id']);
    metaToken = $('meta[name=csrf-token]')[0].content;
    form.elements['authenticity_token'].value = metaToken;
    app.ParticipationLists.resetOptions(event.types);
    return event.label;
  },

  resetOptions: function(types) {
    var select;
    select = $('#role_type');
    if (types.length === 0) {
      select.hide();
      return;
    }
    select.show();
    select.empty();
    return $.each(types, function(_index, type) {
      select.append($('<option></option>').attr('value', type.name).text(type.label));
    });
  }
};
